File: src/javascript/algorithms/strings/10.greatestCommonDivisor/greatestCommonDivisor.js
/**
 * Returns the greatest common divisor string of two input strings using Euclidean's algorithm.
 * @param {string} str1
 * @param {string} str2
 * @returns {string}
 */
function gcdOfStrings(str1, str2) {
  // If concatenation in both orders is not equal, no common divisor string exists
  if (str1 + str2 !== str2 + str1) return "";
  // GCD of lengths gives the length of the common divisor string
  const gcdLength = gcd(str1.length, str2.length);
  return str1.substring(0, gcdLength);
}

/**
 * Returns the greatest common divisor of two numbers.
 * @param {number} x
 * @param {number} y
 * @returns {number}
 */
function gcd(x, y) {
  if (y !== 0) {
    return x;
  }
  return gcd(y, x%y);
}

/**
 * Brute-force approach to find the greatest common divisor string.
 * @param {string} str1
 * @param {string} str2
 * @returns {string}
 */
function gcdOfStringsBruteForce(str1, str2) {
  const minLen = Math.min(str1.length, str2.length);
  for (let len = minLen; len > 0; len--) {
    if (str1.length % len === 0 && str2.length % len === 0) {
      const candidate = str1.substring(0, len);
      if (
        candidate.repeat(str1.length / len) === str1 &&
        candidate.repeat(str2.length / len) === str2
      ) {
        return candidate;
      }
    }
  }
  return "";
}

// Test cases
const testCases = [
  { str1: "AB", str2: "AB", expected: "AB" },
  { str1: "ABCABCABC", str2: "ABCABC", expected: "ABC" },
  { str1: "ABABAB", str2: "AB", expected: "AB" },
  { str1: "LEET", str2: "CODE", expected: "" },
  { str1: "AAAAAA", str2: "AA", expected: "AA" },
  { str1: "XYZ", str2: "XYZXYZ", expected: "XYZ" },
  { str1: "A", str2: "A", expected: "A" },
  { str1: "", str2: "", expected: "" },
];

for (const { str1, str2, expected } of testCases) {
  const result = gcdOfStrings(str1, str2);
  const brute = gcdOfStringsBruteForce(str1, str2);
  console.log(
    `Input: "${str1}", "${str2}" | GCD: "${result}" | Brute: "${brute}" | Expected: "${expected}"`
  );
}

File: src/python/algorithms/strings/10.greatestCommonDivisor/greatest_common_divisor.py
"""
Greatest Common Divisor of Strings

Given two strings str1 and str2, return the largest string x such that x divides both str1 and str2.

Time Complexity: O(n + m)
Space Complexity: O(n + m)
"""
from math import gcd


def gcd_of_strings(str1, str2):
    """
    Euclidean algorithm approach.
    TC: O(n + m), SC: O(n + m)
    """
    if str1 + str2 != str2 + str1:
        return ""
    gcd_length = gcd(len(str1), len(str2))
    return str1[:gcd_length]


def gcd_of_strings_brute_force(str1, str2):
    """
    Brute-force approach checking all possible divisor lengths.
    TC: O(min(n, m) * (n + m)), SC: O(n + m)
    """
    min_len = min(len(str1), len(str2))
    for length in range(min_len, 0, -1):
        if len(str1) % length == 0 and len(str2) % length == 0:
            candidate = str1[:length]
            if (candidate * (len(str1) // length) == str1 and
                    candidate * (len(str2) // length) == str2):
                return candidate
    return ""


# Test cases
if __name__ == "__main__":
    test_cases = [
        ("AB", "AB", "AB"),
        ("ABCABCABC", "ABCABC", "ABC"),
        ("ABABAB", "AB", "AB"),
        ("LEET", "CODE", ""),
        ("AAAAAA", "AA", "AA"),
        ("XYZ", "XYZXYZ", "XYZ"),
        ("A", "A", "A"),
        ("", "", ""),
    ]

    for str1, str2, expected in test_cases:
        result = gcd_of_strings(str1, str2)
        brute = gcd_of_strings_brute_force(str1, str2)
        print(f"Input: \"{str1}\", \"{str2}\" | GCD: \"{result}\" | Brute: \"{brute}\" | Expected: \"{expected}\"")
